grails-app/services/org/pih/warehouse/inventory/CycleCountService.groovy

```diff
@@ -169,4 +173,94 @@ class CycleCountService {
+            if (!cycleCount) {
+                CycleCountDto cycleCountDto = createCycleCount(request, command.facility)
+                cycleCounts.add(cycleCountDto)
+                return
+            }
+            List<CycleCountItem> cycleCountItems = CycleCountItem.findAllByCycleCount(cycleCount)
+            CycleCountDto cycleCountDto = updateCycleCount(cycleCount, cycleCountItems, request)
+            cycleCounts.add(cycleCountDto)
+        }
+        return cycleCounts
+    }
+
+    CycleCountDto createCycleCount(CycleCountStartCommand request, Location facility) {
+        CycleCount newCycleCount = new CycleCount(
+                facility: facility,
+                dateLastRefreshed: new Date()
+        )
+        List<AvailableItem> itemsToSave = determineCycleCountItemsToSave(facility, request.cycleCountRequest.product)
+        newCycleCount.save()
+        // 1:1 association between cycle count and cycle count request
+        request.cycleCountRequest.cycleCount = newCycleCount
+        List<CycleCountItem> cycleCountItems = []
+        itemsToSave.each { AvailableItem availableItem ->
+            CycleCountItem cycleCountItem = new CycleCountItem(
+                    status: CycleCountItemStatus.READY_TO_COUNT,
+                    countIndex: request.countIndex,
+                    quantityOnHand: availableItem.quantityOnHand,
+                    quantityCounted: availableItem.quantityOnHand == 0 ? 0 : null,
+                    cycleCount: newCycleCount,
+                    facility: facility,
+                    inventoryItem: availableItem.inventoryItem,
+                    product: availableItem.inventoryItem.product,
+                    createdBy: AuthService.currentUser,
+                    updatedBy: AuthService.currentUser,
+                    // FIXME: I don't know what default values for custom and draft should be
+                    custom: false,
+                    draft: true,
+            )
+            if (!cycleCountItem.validate()) {
+                throw new ValidationException("Invalid cycle count item", cycleCountItem.errors)
+            }
+            cycleCountItem.save()
+            cycleCountItems.add(cycleCountItem)
+        }
+        return CycleCountDto.createFromCycleCountItems(cycleCountItems)
+    }
+
+    CycleCountDto updateCycleCount(CycleCount cycleCount, List<CycleCountItem> cycleCountItems, CycleCountStartCommand request) {
+        if (cycleCount.status in [CycleCountStatus.TO_REVIEW, CycleCountStatus.CANCELED]) {
+            throw new IllegalArgumentException("Cycle count is already in review or canceled")
+        }
+        if (cycleCount.status in [CycleCountStatus.COUNTING, CycleCountStatus.COUNTED]) {
+            throw new UnsupportedOperationException("Support will be added later")
+        }
+        if (cycleCount.status in [CycleCountStatus.COUNTING, CycleCountStatus.COUNTED, CycleCountStatus.INVESTIGATING, CycleCountStatus.TO_REVIEW]) {
```

Comment:
this was meant to work for cycle count "in progress", so I took this from:
![Screenshot from 2025-01-28 10-11-05](https://github.com/user-attachments/assets/5f12c815-3cf2-4de5-baba-0866aaede3f4)